### PR TITLE
chore: #1677 Rename LetterSoundCorrespondence to LetterSound

### PR DESCRIPTION
**Pull request by SnehaHS65**
This pull request addresses issue #1677 by renaming the class LetterSoundCorrespondence to LetterSound and updating all references throughout the codebase.

### CODE REVIEW

File: src/main/java/ai/elimu/model/content/Word.java [hunk 1]
```diff
@@ -48,12 +48,12 @@ public void setText(String text) {
         this.text = text;
     }
     
-    public List<LetterSoundCorrespondence> getLetterSoundCorrespondences() {
-        return letterSoundCorrespondences;
+    public List<LetterSound> getLetterSoundCorrespondences() {
```
Comment (nya-elimu) [inline src/main/java/ai/elimu/model/content/Word.java:51]:
```suggestion
    public List<LetterSound> getLetterSounds() {
```

**Review by nya-elimu — CHANGES_REQUESTED** [review]

File: src/main/java/ai/elimu/dao/LetterSoundContributionEventDao.java [hunk 2]
```diff
@@ -10,7 +10,7 @@ public interface LetterSoundContributionEventDao extends GenericDao<LetterSoundC
 
     List<LetterSoundCorrespondenceContributionEvent> readAllOrderedByTimeDesc() throws DataAccessException;
 
-    List<LetterSoundCorrespondenceContributionEvent> readAll(LetterSoundCorrespondence letterSound) throws DataAccessException;
+    List<LetterSoundCorrespondenceContributionEvent> readAll(LetterSound letterSound) throws DataAccessException;
```
[the remark this answers is not in the document]
Comment (venkatesh2k3) [reply src/main/java/ai/elimu/dao/LetterSoundContributionEventDao.java:13]: `LetterSoundCorrespondenceContributionEvent` to be changed to `LetterSoundContributionEvent`

File: src/main/java/ai/elimu/dao/jpa/LetterSoundContributionEventDaoJpa.java
Comment (venkatesh2k3) [inline src/main/java/ai/elimu/dao/jpa/LetterSoundContributionEventDaoJpa.java:1]: `LetterSoundCorrespondenceContributionEvent` in SQL query to be changed

File: src/main/java/ai/elimu/dao/jpa/LetterSoundDaoJpa.java
Comment (venkatesh2k3) [inline src/main/java/ai/elimu/dao/jpa/LetterSoundDaoJpa.java:1]: `String letterSoundCorrespondenceLetters` look at the variables declared.

File: src/main/java/ai/elimu/dao/jpa/LetterSoundPeerReviewEventDaoJpa.java
Comment (venkatesh2k3) [inline src/main/java/ai/elimu/dao/jpa/LetterSoundPeerReviewEventDaoJpa.java:1]: Even change the SQL queries.

File: src/main/java/ai/elimu/logic/converters/StringToLetterSoundCorrespondenceConverter.java
Comment (venkatesh2k3) [inline src/main/java/ai/elimu/logic/converters/StringToLetterSoundCorrespondenceConverter.java:1]: `letterSoundCorrespondenceId` change this to `letterSoundId`

File: src/main/java/ai/elimu/model/content/Word.java [hunk 3]
```diff
@@ -48,12 +48,12 @@ public void setText(String text) {
...
+        return letterSounds;
     }
 
-    public void setLetterSoundCorrespondences(List<LetterSoundCorrespondence> letterSoundCorrespondences) {
-        this.letterSoundCorrespondences = letterSoundCorrespondences;
+    public void setLetterSoundCorrespondences(List<LetterSound> letterSounds) {
```
Comment (venkatesh2k3) [inline src/main/java/ai/elimu/model/content/Word.java:55]: Also change setters and getters

File: src/main/java/ai/elimu/model/contributor/LetterSoundCorrespondenceContributionEvent.java [hunk 4]
```diff
@@ -10,13 +10,13 @@ public class LetterSoundCorrespondenceContributionEvent extends ContributionEven
 
     @NotNull
     @ManyToOne
-    private LetterSoundCorrespondence letterSoundCorrespondence;
+    private LetterSound letterSound;
 
-    public LetterSoundCorrespondence getLetterSoundCorrespondence() {
-        return letterSoundCorrespondence;
+    public LetterSound getLetterSoundCorrespondence() {
```
Comment (venkatesh2k3) [inline src/main/java/ai/elimu/model/contributor/LetterSoundCorrespondenceContributionEvent.java:15]: Make sure to change getters and setters.

File: src/main/java/ai/elimu/rest/v2/JpaToGsonConverter.java [hunk 5]
```diff
@@ -129,7 +123,7 @@ public static WordGson getWordGson(Word word) {
             // Word
             wordGson.setText(word.getText());
             List<LetterSoundGson> letterSounds = new ArrayList<>();
-            for (LetterSoundCorrespondence letterSound : word.getLetterSoundCorrespondences()) {
+            for (LetterSound letterSound : word.getLetterSoundCorrespondences()) {
```
Comment (venkatesh2k3) [inline src/main/java/ai/elimu/rest/v2/JpaToGsonConverter.java:126]: getter method to be changed

File: src/main/java/ai/elimu/tasks/LetterSoundCorrespondenceUsageCountScheduler.java [hunk 6]
```diff
@@ -40,28 +40,28 @@ public synchronized void execute() {
         logger.info("words.size(): " + words.size());
         for (Word word : words) {
             logger.info("word.getText(): " + word.getText());
-            for (LetterSoundCorrespondence letterSoundCorrespondence : word.getLetterSoundCorrespondences()) {
-                letterSoundCorrespondenceFrequencyMap.put(letterSoundCorrespondence.getId(),
-                        letterSoundCorrespondenceFrequencyMap.getOrDefault(letterSoundCorrespondence.getId(), 0) + word.getUsageCount());
+            for (LetterSound letterSound : word.getLetterSoundCorrespondences()) {
+                letterSoundCorrespondenceFrequencyMap.put(letterSound.getId(),
+                        letterSoundCorrespondenceFrequencyMap.getOrDefault(letterSound.getId(), 0) + word.getUsageCount());
```
Comment (venkatesh2k3) [inline src/main/java/ai/elimu/tasks/LetterSoundCorrespondenceUsageCountScheduler.java:45]: `letterSoundCorrespondenceFrequencyMap` change it

File: src/main/java/ai/elimu/tasks/SoundUsageCountScheduler.java [hunk 7]
```diff
@@ -42,14 +42,14 @@ public synchronized void execute() {
...
-        // Summarize the usage count of each Word's Sounds based on the LetterSoundCorrespondence's 
+        // Summarize the usage count of each Word's Sounds based on the LetterSound's
         // usage count (see LetterSoundCorrespondenceUsageCountScheduler).
         List<Word> words = wordDao.readAllOrdered();
         logger.info("words.size(): " + words.size());
         for (Word word : words) {
-            for (LetterSoundCorrespondence letterSoundCorrespondence : word.getLetterSoundCorrespondences()) {
-                for (Sound sound : letterSoundCorrespondence.getSounds()) {
-                    soundFrequencyMap.put(sound.getId(), soundFrequencyMap.getOrDefault(sound.getId(), 0) + letterSoundCorrespondence.getUsageCount());
+            for (LetterSound letterSound : word.getLetterSoundCorrespondences()) {
```
Comment (venkatesh2k3) [inline src/main/java/ai/elimu/tasks/SoundUsageCountScheduler.java:50]: `getLetterSoundCorrespondences()`

File: src/main/java/ai/elimu/util/csv/CsvContentExtractionHelper.java [hunk 8]
```diff
@@ -56,10 +53,10 @@ public class CsvContentExtractionHelper {
     /**
      * For information on how the CSV files were generated, see {@link LetterSoundCsvExportController#handleRequest}.
      */
-    public static List<LetterSoundCorrespondence> getLetterSoundCorrespondencesFromCsvBackup(File csvFile, LetterDao letterDao, SoundDao soundDao, LetterSoundDao letterSoundDao) {
+    public static List<LetterSound> getLetterSoundCorrespondencesFromCsvBackup(File csvFile, LetterDao letterDao, SoundDao soundDao, LetterSoundDao letterSoundDao) {
         logger.info("getLetterSoundCorrespondencesFromCsvBackup");
```
Comment (venkatesh2k3) [inline src/main/java/ai/elimu/util/csv/CsvContentExtractionHelper.java:57]: `getLetterSoundCorrespondencesFromCsvBackup`

**Review by venkatesh2k3 — CHANGES_REQUESTED** [review]
Getter and Setter methods to be changed and variables declared, The method return type to be changed in each file

**Comment by SnehaHS65** [comment]
@venkatesh2k3 , I have changed in all the java files as you mentioned, should I change in webapp folder(in html and css files) as well? Let me know and I can commit

**Review by venkatesh2k3 — DISMISSED** [review]

**Review by jo-elimu — CHANGES_REQUESTED** [review]
@SnehaHS65 Thank you for opening this pull request 🙂

I will have to ask you to revert the second commit in your pull request: https://github.com/elimu-ai/webapp/pull/1682/commits/35093efec841f6c7b2b8233c732bc090ce548afc

This is because:

- The scope for the changes should only be for `LetterSoundCorrespondence.java`, and not for additional entities. (The number of code changes is already high in the initial commit, so we shouldn't add additional complexity.)
- The changes are touching the database layer, which means that we will also have to create a database migration script for your changes. See [DB_MIGRATION.md](https://github.com/elimu-ai/webapp/blob/main/DB_MIGRATION.md).
- A database schema export file is missing. You can generate it by running `mvn jetty:run` to launch the webapp locally. You will then see code changes in the `jpa-schema-export.sql` file, and those changes should also be included in your PR.
    

> [!NOTE]
> The reason why we need to create a DB migration script is because we already have a MySQL production database. So anytime we make changes to the database structure, we have to write SQL scripts for updating the database. See [DB_MIGRATION.md](https://github.com/elimu-ai/webapp/blob/main/DB_MIGRATION.md).

> [!TIP]
> When you create the database migration script, you can look at the changes in `jpa-schema-export.sql` to know which tables/columns to modify.

As this PR is already quite large, you might want to split it up into smaller parts to make sure that it doesn't break the database structure. For example one PR for only renaming the `LetterSoundCorrespondence` database table, and another PR for renaming  the `LetterSoundCorrespondence_id` references another database table, etc.

As you can see in the `jpa-schema-export.sql` diff, it gets quite complicated with all the changes to the database column names, primary keys, and foreign key constraints. And that is why I recommend we reduce the size of your PR.

**Comment by SnehaHS65** [comment]
@jo-elimu , Can you review the latest commit now, I refactored `LetterSoundCorrespondence.java` to `LetterSound.java`, ran this command `mvn jetty:run`, included `jpa-schema-export.sql` in the PR and also added DB script for the same.

**Comment by SnehaHS65** [comment]
Hi @jo-elimu , @nya-elimu , @venkatesh2k3 , can you please look at the latest changes that I did to rename LetterSoundCorrespondence to LetterSound. This is the first PR to just rename the LetterSoundCorrespondence database table as suggested by @jo-elimu. If you approve this, I can raise one more PR for renaming the LetterSoundCorrespondence_id references.

**Comment by jo-elimu** [comment]
> Hi @jo-elimu , @nya-elimu , @venkatesh2k3 , can you please look at the latest changes that I did to rename LetterSoundCorrespondence to LetterSound. This is the first PR to just rename the LetterSoundCorrespondence database table as suggested by @jo-elimu. If you approve this, I can raise one more PR for renaming the LetterSoundCorrespondence_id references.

@SnehaHS65 If you look at your changes in [`jpa-schema-export.sql`](https://github.com/elimu-ai/webapp/pull/1682/files#diff-94be1c5456b98ad8b18c1bf1dfb56ed60c79d2c490d5c12d93a1d99dc77fc48d), you'll see that it contains around 16 changes, and these 16 changes would also have to be included in the DB migration script.

This is why I suggested above that we split #1677 into smaller pull requests to make each database migration less complicated.

I've created an example here: https://github.com/elimu-ai/webapp/pull/1713. This pull request only makes one small change (renaming from `Word#letterSoundCorrespondences` to `Word#letterSoundCorrespondences`). After merging that PR, we can perform a release and check if the migration script runs without errors. If so, we can move on to another small part of the issue.

**Comment by SnehaHS65** [comment]
> > Hi @jo-elimu , @nya-elimu , @venkatesh2k3 , can you please look at the latest changes that I did to rename LetterSoundCorrespondence to LetterSound. This is the first PR to just rename the LetterSoundCorrespondence database table as suggested by @jo-elimu. If you approve this, I can raise one more PR for renaming the LetterSoundCorrespondence_id references.
> 
> @SnehaHS65 If you look at your changes in [`jpa-schema-export.sql`](https://github.com/elimu-ai/webapp/pull/1682/files#diff-94be1c5456b98ad8b18c1bf1dfb56ed60c79d2c490d5c12d93a1d99dc77fc48d), you'll see that it contains around 16 changes, and these 16 changes would also have to be included in the DB migration script.
> 
> This is why I suggested above that we split #1677 into smaller pull requests to make each database migration less complicated.
> 
> I've created an example here: #1713. This pull request only makes one small change (renaming from `Word#letterSoundCorrespondences` to `Word#letterSounds`). After merging that PR, we can perform a release and verify that the migration script runs without errors. If so, we can move on to another small part of the issue.

@jo-elimu ,So there are ContributionEvent and PeerReviewEvent which uses this letterSoundCorrespondence, I will now only edit for ContributionEvent first and commit?

**Comment by jo-elimu** [comment]
> @jo-elimu ,So there are ContributionEvent and PeerReviewEvent which uses this letterSoundCorrespondence, I will now only edit for ContributionEvent first and commit?

@SnehaHS65 Yes, that's right :-)

I suggest you close this pull request, and open a new pull request for renaming the `LetterSoundCorrespondenceContributionEvent` class. And remember to include the DB migration script as well (see similar example in [`2004010.sql`](https://github.com/elimu-ai/webapp/blob/main/src/main/resources/db/migration/2004010.sql)).

**Pull request closed without merging**